corehq/apps/export/tests/test_export_data_schema.py

```diff
@@ -0,0 +1,162 @@
+
+    def _get_merged_schema(self, form_name1, form_name2):
+        form_xml = self.get_xml(form_name1)
+        form_xml2 = self.get_xml(form_name2)
+        schema = ExportDataSchema._generate_schema_from_xform(
+            XForm(form_xml),
+            ['en'],
+            1
+        )
+        schema2 = ExportDataSchema._generate_schema_from_xform(
+            XForm(form_xml2),
+            ['en'],
+            2
+        )
+
+        return ExportDataSchema._merge_schema(schema, schema2)
+
+    def test_simple_merge(self):
+        """Tests merging of a form that adds a question to the form"""
+        merged = self._get_merged_schema('basic_form', 'basic_form_version2')
+
+        self.assertEqual(len(merged.group_schemas), 1)
+
+        group_schema = merged.group_schemas[0]
+        self.assertEqual(len(group_schema.items), 3)
+        self.assertTrue(all(map(lambda item: item.last_occurrence == 2, group_schema.items)))
+
+    def test_merge_deleted(self):
+        """Tests merging of a form that deletes a question from its form"""
+        merged = self._get_merged_schema('basic_form', 'basic_form_version2_delete')
+
+        self.assertEqual(len(merged.group_schemas), 1)
+
+        group_schema = merged.group_schemas[0]
+        self.assertEqual(len(group_schema.items), 2)
+
+        v1items = filter(lambda item: item.last_occurrence == 1, group_schema.items)
+        v2items = filter(lambda item: item.last_occurrence == 2, group_schema.items)
+
+        self.assertEqual(
+            len(v2items),
+            1,
+            'There should be 1 item that was found in the second version. There was {}'.format(len(v2items))
+        )
+        self.assertEqual(
+            len(v1items),
+            1,
+            'There should be 1 item that was found in the first version. There was {}'.format(len(v1items))
+        )
+
+    def test_multiple_choice_merge(self):
+        """Tests merging of a form that changes the options to a multiple choice question"""
+        merged = self._get_merged_schema('multiple_choice_form', 'multiple_choice_form_version2')
+
+        self.assertEqual(len(merged.group_schemas), 1)
+
+        group_schema = merged.group_schemas[0]
+        self.assertEqual(len(group_schema.items), 2)
+
+        v2items = filter(lambda item: item.last_occurrence == 2, group_schema.items)
+        self.assertEqual(
+            len(v2items),
+            2,
+        )
+
+        multichoice = filter(lambda item: item.path == ['data', 'question2'], group_schema.items)[0]
+        self.assertEqual(len(multichoice.options), 3)
+        self.assertEqual(
+            len(filter(lambda o: o.last_occurrence == 2, multichoice.options)),
+            2,
+        )
+
+        self.assertEqual(
+            len(filter(lambda o: o.last_occurrence == 1, multichoice.options)),
+            1,
+        )
+
+    def test_merge_repeat_group_changed_id(self):
+        """This tests merging forms that change a question to a repeat group"""
```

Comment:
i'm so excited for this. people can change there form in any way and it'll "Just Work"